src/hotspot/share/utilities/istream.hpp

```diff
@@ -0,0 +1,447 @@
+  void set_done();
+
+  // Reports if this stream has had an error was reported on it.
+  bool error() const {
+    return _input_state == ERR_STATE;
+  }
+
+  // Set this stream done with an error, if the argument is true.
+  // If it is false but there is an error condition, clear the error.
+  // Otherwise do nothing.
+  void set_error(bool error_condition = true);
+
+  // lineno is the 1-based ordinal of the current line; it starts at one
+  size_t lineno() const         { preload(); return _line_count; }
+
+  // Copy the current line to the given output stream.
+  void print_on(outputStream* out);
+
+  // Copy the current line to the given output stream, and also call cr().
+  void print_cr_on(outputStream* out) {
+    print_on(out); out->cr();
+  }
+
+#ifdef ASSERT
+  void dump(const char* what = nullptr);
+  static int coverage_mode(int mode, int& cases, int& total, int& zeroes);
+#else
+  void dump(const char* what = nullptr) { }
+#endif
+
+
+  // Block-oriented input, which treats all bytes equally.
+  class Input : public CHeapObjBase {
+  public:
+    // Read some characters from an external source into the line buffer.
+    // If there are no more, return zero, otherwise return non-zero.
+    // It must be OK to call read even after it returns zero.
+    virtual size_t read(char* buf, size_t size) = 0;
+    // Example: read(b,s) { return fread(b, 1, s, _my_fp); }
+    // Example: read(b,s) { return 0; } // never more than the initial buffer
+
+    // Give the current number of bytes already produced by the source.
+    // Give (size_t)-1 if this source does have a tracked position.
+    // A tracked position increments by the result of every call to read.
+    virtual size_t position() { return -1; }
+
+    // Give the remaining number of bytes which might be produced in the future.
+    // Give (size_t)-1 if this source does not keep track of that number.
+    virtual size_t remaining() { return -1; }
+
+    // Rewind so that the position appears to be the given one.
+    // Return the new position, or else (size_t)-1 if the request fails.
+    virtual size_t set_position(size_t position) { return -1; }
+
+    // If it is backed by a resource that needs closing, do so.
+    virtual void close() { }
+  };
+};
+
+template<typename BlockClass>
+class BlockInputStream : public inputStream {
+  BlockClass _input;
+ public:
+  template<typename... Arg>
+  BlockInputStream(Arg... arg)
+    : _input(arg...) {
+    set_input(&_input);
+  }
+};
+
+// for reading lines from files
+class FileInput : public inputStream::Input {
+  NONCOPYABLE(FileInput);
+
+ protected:
+  fileStream& _fs;
+  fileStream _private_fs;
+
+  // it does not seem likely there are such file streams around
+  FileInput(fileStream& fs)
+    : _fs(fs)
+  { }
+
+ public:
+  // just forward all the constructor arguments to the wrapped line-input class
+  template<typename... Arg>
+  FileInput(Arg... arg)
+    : _fs(_private_fs), _private_fs(arg...)
+  {
+  }
```

Comment:
can you put  {} on line 372, this doesn't need so much vertical space.

Reply:
Is this ... in template typename parameters allowed by our coding standard (whatever it means?)

Reply:
Yes.  It’s called “variadic templates”; see `doc/hotspot-style.md`.   The usage appears twice already, in `metaprogramming/logical.hpp` and `asm/register.hpp`.  The usage here is pretty straightforward IMO:  It just forwards constructor argument lists unchanged.  This is useful for wrappers (where you construct the wrapped thing as a field) or subclass factoring (where you construct the super).